pkg/controller/bootstrap/bootstrapsigner.go

```diff
@@ -0,0 +1,269 @@
+	ConfigMapResync time.Duration
+
+	// SecretResync is the time.Duration at which to fully re-list secrets.
+	// If zero, re-list will be delayed as long as possible
+	SecretResync time.Duration
+}
+
+// DefaultBootstrapSignerOptions returns a set of default options for creating a
+// BootstrapSigner
+func DefaultBootstrapSignerOptions() BootstrapSignerOptions {
+	return BootstrapSignerOptions{
+		ConfigMapNamespace: api.NamespacePublic,
+		ConfigMapName:      configMapClusterInfo,
+	}
+}
+
+// BootstrapSigner is a controller that signs a ConfigMap with a set of tokens.
+type BootstrapSigner struct {
+	client       clientset.Interface
+	configMapKey string
+
+	configMaps cache.Store
+	secrets    cache.Store
+
+	// Since we join two objects, we'll watch both of them with controllers.
+	configMapsController *cache.Controller
+	secretsController    *cache.Controller
+}
+
+// NewBootstrapSigner returns a new *BootstrapSigner.
+func NewBootstrapSigner(cl clientset.Interface, options BootstrapSignerOptions) *BootstrapSigner {
+	e := &BootstrapSigner{
+		client:       cl,
+		configMapKey: options.ConfigMapNamespace + "/" + options.ConfigMapName,
+	}
+	if cl.Core().RESTClient().GetRateLimiter() != nil {
+		metrics.RegisterMetricAndTrackRateLimiterUsage("bootstrap_signer", cl.Core().RESTClient().GetRateLimiter())
+	}
+	configMapSelector := fields.SelectorFromSet(map[string]string{api.ObjectNameField: options.ConfigMapName})
+	e.configMaps, e.configMapsController = cache.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(lo v1.ListOptions) (runtime.Object, error) {
+				lo.FieldSelector = configMapSelector.String()
+				return e.client.Core().ConfigMaps(options.ConfigMapNamespace).List(lo)
+			},
+			WatchFunc: func(lo v1.ListOptions) (watch.Interface, error) {
+				lo.FieldSelector = configMapSelector.String()
+				return e.client.Core().ConfigMaps(options.ConfigMapNamespace).Watch(lo)
+			},
+		},
+		&v1.ConfigMap{},
+		options.ConfigMapResync,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(_ interface{}) { e.signConfigMap() },
+			UpdateFunc: func(_, _ interface{}) { e.signConfigMap() },
+		},
+	)
+
+	secretSelector := fields.SelectorFromSet(map[string]string{api.SecretTypeField: string(api.SecretTypeBootstrapToken)})
+	e.secrets, e.secretsController = cache.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(lo v1.ListOptions) (runtime.Object, error) {
+				lo.FieldSelector = secretSelector.String()
+				return e.client.Core().Secrets(api.NamespaceSystem).List(lo)
+			},
+			WatchFunc: func(lo v1.ListOptions) (watch.Interface, error) {
+				lo.FieldSelector = secretSelector.String()
+				return e.client.Core().Secrets(api.NamespaceSystem).Watch(lo)
+			},
+		},
+		&v1.Secret{},
+		options.SecretResync,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(_ interface{}) { e.signConfigMap() },
+			UpdateFunc: func(_, _ interface{}) { e.signConfigMap() },
+			DeleteFunc: func(_ interface{}) { e.signConfigMap() },
```

Comment:
Also, a full queue here isn't necessary.  I just want to make sure that I "poke" my sign function in a serialized way and queue up a "re-poke" if it is currently running.  I can't find an existing helper for this so I'm going to use a simple `workqueue.queue` with a single static item key.  I figure reusing that is better than writing more subtle multi-threaded code.